pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -61,27 +61,29 @@ public class ModelInterpreter implements Serializable {
 
         if (root != null) {
             // Attach the stages model to the run for introspection etc.
-            Utils.attachDeclarativeActions(script)
+            root = Utils.attachDeclarativeActions(root, script)
             boolean postBuildRun = false
 
             try {
                 executeProperties(root)
 
-                // Entire build, including notifications, runs in the withEnv.
-                withEnvBlock(root.getEnvVars()) {
-                    inWrappers(root.options) {
-                        // Stage execution and post-build actions run in try/catch blocks, so we still run post-build actions
-                        // even if the build fails.
-                        // We save the caught error, if any, for throwing at the end of the build.
-                        inDeclarativeAgent(root, root.agent) {
+                // Entire build, including notifications, runs in the agent.
+                inDeclarativeAgent(root, root.agent) {
```

Comment:
So you can no longer use environment variables to declare agent? Sort of hinders one of the use cases of JENKINS-41759 where the properties would be the marker file and declare what docker image to run in.